File: apps/android/core/model/src/main/java/com/pokr/android/core/model/PlayedHand.kt
package com.pokr.android.core.model

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

private val RankChar = mapOf(
    14 to "A",
    13 to "K",
    12 to "Q",
    11 to "J",
    10 to "T",
    9 to "9",
    8 to "8",
    7 to "7",
    6 to "6",
    5 to "5",
    4 to "4",
    3 to "3",
    2 to "2",
)

private val LenientJson = Json {
    ignoreUnknownKeys = true
    isLenient = true
}

data class PlayedHandLevel(
    val id: String,
    val handId: String,
    val source: String,
    val startedAt: Long,
    val holeCards: Pair<String, String>?,
    val community: List<String>,
    val won: Boolean,
    val winnerName: String?,
    val handName: String?,
)

fun cardCode(raw: JsonElement?): String? {
    if (raw == null) return null
    when (raw) {
        is JsonPrimitive -> {
            val s = raw.contentOrNull ?: return null
            return if (s.length >= 2) s else null
        }
        is JsonObject -> {
            val suit = raw["suit"]?.jsonPrimitive?.contentOrNull ?: return null
            val rankEl = raw["rank"] ?: return null
            val rankPrim = rankEl.jsonPrimitive
            val rankNum = rankPrim.intOrNull
            if (rankNum != null) {
                val ch = RankChar[rankNum] ?: return null
                return "$ch$suit"
            }
            val rankStr = rankPrim.contentOrNull ?: return null
            return "$rankStr$suit"
        }
        else -> return null
    }
}

fun cardCodes(raw: JsonElement?): List<String> {
    val arr = raw as? JsonArray ?: return emptyList()
    return arr.mapNotNull { cardCode(it) }
}

fun parsePlayedHand(row: MyHandRow, userId: String): PlayedHandLevel {
    val result = parseResult(row.resultJson)
    val players = result?.get("players")?.jsonArrayOrNull().orEmpty()
    var seat: Int? = null
    var hole: List<String> = emptyList()
    for (p in players) {
        val rec = p as? JsonObject ?: continue
        if (rec["userId"]?.jsonPrimitive?.contentOrNull != userId) continue
        seat = rec["seat"]?.jsonPrimitive?.intOrNull
        hole = cardCodes(rec["holeCards"])
        break
    }
    val winnersRaw = result?.get("winners")?.jsonArrayOrNull().orEmpty()
    var won = false
    var winnerName: String? = null
    var handName: String? = null
    for (w in winnersRaw) {
        val rec = w as? JsonObject ?: continue
        val wSeat = rec["seat"]?.jsonPrimitive?.intOrNull
        if (wSeat != null && wSeat == seat) {
            won = true
            handName = rec["handName"]?.jsonPrimitive?.contentOrNull
        }
        if (winnerName == null) {
            winnerName = rec["name"]?.jsonPrimitive?.contentOrNull?.takeIf { it.isNotBlank() }
        }
    }
    val pair = if (hole.size >= 2) hole[0] to hole[1] else null
    return PlayedHandLevel(
        id = row.id,
        handId = row.handId,
        source = row.source.ifBlank { "online" },
        startedAt = startedAtMs(row.startedAt),
        holeCards = pair,
        community = cardCodes(result?.get("community")),
        won = won,
        winnerName = winnerName,
        handName = handName,
    )
}

/** Newest fetched hand maps to [handsPlayed], then count backward. */
fun handsByLevel(hands: List<PlayedHandLevel>, handsPlayed: Int): Map<Int, PlayedHandLevel> {
    val map = LinkedHashMap<Int, PlayedHandLevel>()
    hands.forEachIndexed { i, hand ->
        val level = handsPlayed - i
        if (level >= 1) map[level] = hand
    }
    return map
}

fun formatHandWhen(ms: Long): String {
    if (ms <= 0L) return ""
    return try {
        val fmt = SimpleDateFormat("MMM d, h:mm a", Locale.getDefault())
        fmt.timeZone = TimeZone.getDefault()
        fmt.format(Date(ms))
    } catch (_: Exception) {
        ""
    }
}

private fun parseResult(resultJson: String): JsonObject? {
    if (resultJson.isBlank()) return null
    return try {
        when (val parsed = LenientJson.parseToJsonElement(resultJson)) {
            is JsonObject -> parsed
            else -> null
        }
    } catch (_: Exception) {
        null
    }
}

private fun startedAtMs(raw: String): Long {
    if (raw.isBlank()) return 0L
    raw.toLongOrNull()?.let { return it }
    return try {
        // ISO-8601 from Nest / Postgres
        java.time.Instant.parse(raw).toEpochMilli()
    } catch (_: Exception) {
        try {
            val withMillis = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
            withMillis.parse(raw)?.time ?: 0L
        } catch (_: Exception) {
            0L
        }
    }
}

private fun JsonElement.jsonArrayOrNull(): JsonArray? = this as? JsonArray

File: apps/web/features/progress/playedHand.ts
import type { MyHandRow } from '@/lib/api';

const RANK_CHAR: Record<number, string> = {
  14: 'A',
  13: 'K',
  12: 'Q',
  11: 'J',
  10: 'T',
  9: '9',
  8: '8',
  7: '7',
  6: '6',
  5: '5',
  4: '4',
  3: '3',
  2: '2',
};

export function cardCode(raw: unknown): string | null {
  if (typeof raw === 'string' && raw.length >= 2) return raw;
  if (!raw || typeof raw !== 'object') return null;
  const rec = raw as { rank?: unknown; suit?: unknown };
  const suit = typeof rec.suit === 'string' ? rec.suit : null;
  if (!suit) return null;
  if (typeof rec.rank === 'number' && RANK_CHAR[rec.rank]) {
    return `${RANK_CHAR[rec.rank]}${suit}`;
  }
  if (typeof rec.rank === 'string') return `${rec.rank}${suit}`;
  return null;
}

export function cardCodes(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  return raw.map(cardCode).filter((c): c is string => !!c);
}

export type ShownPlayerHand = {
  userId: string;
  name: string;
  holeCards: [string, string] | null;
  isViewer: boolean;
  isWinner: boolean;
};

export type PlayedHandLevel = {
  id: string;
  handId: string;
  source: string;
  startedAt: number;
  holeCards: [string, string] | null;
  community: string[];
  won: boolean;
  winnerName: string | null;
  handName: string | null;
  /** Viewer + anyone whose hole cards were revealed at showdown. */
  shownPlayers: ShownPlayerHand[];
};

function parseResult(resultJson: unknown): Record<string, unknown> | null {
  if (resultJson && typeof resultJson === 'object') {
    return resultJson as Record<string, unknown>;
  }
  if (typeof resultJson !== 'string') return null;
  try {
    const parsed = JSON.parse(resultJson) as unknown;
    return parsed && typeof parsed === 'object' ? (parsed as Record<string, unknown>) : null;
  } catch {
    return null;
  }
}

function startedAtMs(raw: string | number | Date | null | undefined): number {
  if (typeof raw === 'number' && Number.isFinite(raw)) return raw;
  if (raw instanceof Date) return raw.getTime();
  if (typeof raw === 'string') {
    const n = Date.parse(raw);
    return Number.isNaN(n) ? 0 : n;
  }
  return 0;
}

export function parsePlayedHand(row: MyHandRow, userId: string): PlayedHandLevel {
  const result = parseResult(row.resultJson);
  const players = Array.isArray(result?.players) ? result.players : [];
  const nameBySeat = new Map<number, string>();
  const playerBySeat = new Map<
    number,
    { userId: string; name: string; holeCards: string[]; revealed: boolean }
  >();

  let seat: number | null = null;
  let hole: string[] = [];
  for (const p of players) {
    if (!p || typeof p !== 'object') continue;
    const rec = p as {
      userId?: unknown;
      seat?: unknown;
      holeCards?: unknown;
      name?: unknown;
      revealed?: unknown;
    };
    const pid = typeof rec.userId === 'string' ? rec.userId : '';
    const pname =
      typeof rec.name === 'string' && rec.name
        ? rec.name
        : pid
          ? pid
          : 'Player';
    const cards = cardCodes(rec.holeCards);
    if (typeof rec.seat === 'number') {
      nameBySeat.set(rec.seat, pname);
      playerBySeat.set(rec.seat, {
        userId: pid,
        name: pname,
        holeCards: cards,
        revealed: rec.revealed === true,
      });
    }
    if (pid === userId) {
      if (typeof rec.seat === 'number') seat = rec.seat;
      hole = cards;
    }
  }

  const winnersRaw = Array.isArray(result?.winners) ? result.winners : [];
  let won = false;
  let handName: string | null = null;
  const winnerNames: string[] = [];
  const winnerSeats = new Set<number>();
  for (const w of winnersRaw) {
    if (!w || typeof w !== 'object') continue;
    const rec = w as { seat?: unknown; name?: unknown; handName?: unknown };
    if (typeof rec.seat !== 'number') continue;
    winnerSeats.add(rec.seat);
    const resolved =
      (typeof rec.name === 'string' && rec.name) || nameBySeat.get(rec.seat) || null;
    if (resolved && !winnerNames.includes(resolved)) winnerNames.push(resolved);
    if (!handName && typeof rec.handName === 'string' && rec.handName) {
      handName = rec.handName;
    }
    if (rec.seat === seat) won = true;
  }
  const winnerName = winnerNames.length > 0 ? winnerNames.join(' & ') : null;
  const pair: [string, string] | null = hole.length >= 2 ? [hole[0]!, hole[1]!] : null;

  const shownPlayers: ShownPlayerHand[] = [];
  const seen = new Set<string>();
  for (const [seatNum, p] of playerBySeat) {
    const isViewer = p.userId === userId;
    const hasCards = p.holeCards.length >= 2;
    // Viewer always; others only when hole cards survived redaction (revealed at showdown).
    if (!isViewer && !hasCards) continue;
    const key = p.userId || `${seatNum}:${p.name}`;
    if (seen.has(key)) continue;
    seen.add(key);
    shownPlayers.push({
      userId: p.userId,
      name: p.name,
      holeCards: hasCards ? [p.holeCards[0]!, p.holeCards[1]!] : isViewer ? pair : null,
      isViewer,
      isWinner: winnerSeats.has(seatNum),
    });
  }
  // Viewer first, then winners, then others by name.
  shownPlayers.sort((a, b) => {
    if (a.isViewer !== b.isViewer) return a.isViewer ? -1 : 1;
    if (a.isWinner !== b.isWinner) return a.isWinner ? -1 : 1;
    return a.name.localeCompare(b.name);
  });

  return {
    id: row.id,
    handId: row.handId,
    source: row.source ?? 'online',
    startedAt: startedAtMs(row.startedAt),
    holeCards: pair,
    community: cardCodes(result?.community),
    won,
    winnerName,
    handName,
    shownPlayers,
  };
}

/** Newest fetched hand maps to `handsPlayed`, then count backward. */
export function handsByLevel(
  hands: PlayedHandLevel[],
  handsPlayed: number,
): Map<number, PlayedHandLevel> {
  const map = new Map<number, PlayedHandLevel>();
  hands.forEach((hand, i) => {
    const level = handsPlayed - i;
    if (level >= 1) map.set(level, hand);
  });
  return map;
}

export function formatHandWhen(ms: number): string {
  if (!ms) return '';
  const d = new Date(ms);
  if (Number.isNaN(d.getTime())) return '';
  return d.toLocaleString(undefined, {
    month: 'short',
    day: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
  });
}
